File: todoforai_edge/handlers/docx_handler.py
import os
import zipfile
import shutil
import xml.dom.minidom
import xml.etree.ElementTree as ET
import logging
import json

logger = logging.getLogger("todoforai-edge")

def is_valid_xml(content):
    """Check if content is valid XML or valid multi-file format containing XML"""
    return True # now we don't check, just save even if it is not valid

def extract_docx_content(docx_path):
    """Extract readable content from DOCX file"""
    try:
        with zipfile.ZipFile(docx_path, 'r') as zip_file:
            # Extract main document XML
            xml_content = zip_file.read("word/document.xml").decode('utf-8')
            formatted_xml = xml.dom.minidom.parseString(xml_content).toprettyxml()
            
            # Return plain XML for DOCX
            return formatted_xml
    except Exception as e:
        raise ValueError(f"Failed to extract DOCX content: {str(e)}")

def extract_xlsx_content(xlsx_path):
    """Extract readable content from Excel file as multi-file format with clear delimiters"""
    try:
        with zipfile.ZipFile(xlsx_path, 'r') as zip_file:
            result = {}
            
            # Extract worksheet(s)
            worksheet_files = [f for f in zip_file.namelist() if f.startswith("xl/worksheets/") and f.endswith(".xml")]
            for worksheet_path in worksheet_files:
                xml_content = zip_file.read(worksheet_path).decode('utf-8')
                formatted_xml = xml.dom.minidom.parseString(xml_content).toprettyxml()
                key = worksheet_path.replace("xl/", "")
                result[key] = formatted_xml
            
            # Extract shared strings if exists
            if "xl/sharedStrings.xml" in zip_file.namelist():
                xml_content = zip_file.read("xl/sharedStrings.xml").decode('utf-8')
                formatted_xml = xml.dom.minidom.parseString(xml_content).toprettyxml()
                result["sharedStrings.xml"] = formatted_xml
            
            # Extract styles if exists
            if "xl/styles.xml" in zip_file.namelist():
                xml_content = zip_file.read("xl/styles.xml").decode('utf-8')
                formatted_xml = xml.dom.minidom.parseString(xml_content).toprettyxml()
                result["styles.xml"] = formatted_xml
            
            # Dump to multi-file delimited format
            return dump_multi_file_content(result)
            
    except Exception as e:
        raise ValueError(f"Failed to extract Excel content: {str(e)}")

def save_docx_content(docx_path, xml_content):
    """Save XML content back to DOCX file by updating the document.xml inside the ZIP"""
    try:
        # Read the existing DOCX file
        with zipfile.ZipFile(docx_path, 'r') as zip_read:
            # Get all files in the ZIP
            file_list = zip_read.namelist()
            if 'word/document.xml' not in file_list:
                raise ValueError("Invalid DOCX: 'word/document.xml' not found")
            
            # Create a new DOCX file with updated content
            temp_path = docx_path + '.tmp'
            with zipfile.ZipFile(temp_path, 'w', zipfile.ZIP_DEFLATED) as zip_write:
                # Copy all files except document.xml
                for file_name in file_list:
                    if file_name != 'word/document.xml':
                        zip_write.writestr(file_name, zip_read.read(file_name))
                
                # Write the new document.xml content
                clean_xml = _clean_header_from_xml(xml_content)
                zip_write.writestr('word/document.xml', clean_xml.encode('utf-8'))
        
        # Replace the original file with the updated one
        shutil.move(temp_path, docx_path)
        
    except Exception as e:
        # Clean up temp file if it exists
        if os.path.exists(docx_path + '.tmp'):
            os.remove(docx_path + '.tmp')
        raise ValueError(f"Failed to save DOCX content: {str(e)}")

def save_xlsx_content(xlsx_path, multi_file_content):
    """Save multi-file content back to Excel file by parsing delimited format"""
    try:
        # Parse the multi-file format
        xml_files = parse_multi_file_content(multi_file_content)
        
        # Read the existing Excel file
        with zipfile.ZipFile(xlsx_path, 'r') as zip_read:
            file_list = zip_read.namelist()
            
            # Create a new Excel file with updated content
            temp_path = xlsx_path + '.tmp'
            with zipfile.ZipFile(temp_path, 'w', zipfile.ZIP_DEFLATED) as zip_write:
                # Copy all files except the ones we're updating
                files_to_update = set()
                for key in xml_files.keys():
                    full_path = f"xl/{key}"
                    files_to_update.add(full_path)
                
                for file_name in file_list:
                    if file_name not in files_to_update:
                        zip_write.writestr(file_name, zip_read.read(file_name))
                
                # Write the updated XML files
                for key, xml_content in xml_files.items():
                    full_path = f"xl/{key}"
                    clean_xml = _clean_header_from_xml(xml_content)
                    zip_write.writestr(full_path, clean_xml.encode('utf-8'))
        
        # Replace the original file with the updated one
        shutil.move(temp_path, xlsx_path)
        
    except Exception as e:
        # Clean up temp file if it exists
        if os.path.exists(xlsx_path + '.tmp'):
            os.remove(xlsx_path + '.tmp')
        raise ValueError(f"Failed to save Excel content: {str(e)}")

def parse_multi_file_content(content):
    """Parse multi-file content format into a dictionary of filename -> xml_content"""
    result = {}
    parts = content.split("=== FILE: ")
    for part in parts[1:]:
        if " ===" not in part:
            continue
        filename_end = part.find(" ===")
        if filename_end == -1:
            continue
        filename = part[:filename_end]
        xml_content = part[filename_end + 4:].strip()
        if filename and xml_content:
            result[filename] = xml_content
    return result

def dump_multi_file_content(files_map):
    """Dump a dict of filename -> xml_content into the delimited multi-file format"""
    parts = []
    for key, content in files_map.items():
        parts.append(f"=== FILE: {key} ===")
        parts.append(content.strip())
    # Join with double newlines to match extract_xlsx_content format
    return "\n\n".join(parts)

def _clean_header_from_xml(xml_content, header_marker=None):
    """Remove the header comments from XML content"""
    clean_xml = xml_content
    if header_marker and clean_xml.startswith(header_marker):
        lines = clean_xml.split('\n')
        xml_start = 0
        for i, line in enumerate(lines):
            if line.strip().startswith('<?xml'):
                xml_start = i
                break
        clean_xml = '\n'.join(lines[xml_start:])
    else:
        lines = clean_xml.split('\n')
        for i, line in enumerate(lines):
            if line.strip().startswith('<?xml'):
                clean_xml = '\n'.join(lines[i:])
                break
    return clean_xml

File: bun/src/docx-handler.ts
import fs from "fs";
import { unzipSync, zipSync, type Unzipped } from "fflate";

// ── Pretty-print XML (simple regex-based indenter) ──

function prettyPrintXml(xml: string): string {
  // Normalize to single line first
  let formatted = "";
  let indent = 0;
  const parts = xml.replace(/>\s*</g, "><").split(/(<[^>]+>)/);
  for (const part of parts) {
    if (!part.trim()) continue;
    if (part.startsWith("</")) {
      indent = Math.max(indent - 1, 0);
      formatted += "  ".repeat(indent) + part + "\n";
    } else if (part.startsWith("<?")) {
      formatted += part + "\n";
    } else if (part.startsWith("<") && !part.endsWith("/>") && !part.startsWith("<!")) {
      formatted += "  ".repeat(indent) + part + "\n";
      indent++;
    } else if (part.endsWith("/>")) {
      formatted += "  ".repeat(indent) + part + "\n";
    } else {
      formatted += "  ".repeat(indent) + part + "\n";
    }
  }
  return formatted;
}

// ── Multi-file format helpers ──

export function parseMultiFileContent(content: string): Record<string, string> {
  const result: Record<string, string> = {};
  const parts = content.split("=== FILE: ");
  for (const part of parts.slice(1)) {
    const endIdx = part.indexOf(" ===");
    if (endIdx === -1) continue;
    const filename = part.slice(0, endIdx);
    const xmlContent = part.slice(endIdx + 4).trim();
    if (filename && xmlContent) result[filename] = xmlContent;
  }
  return result;
}

export function dumpMultiFileContent(filesMap: Record<string, string>): string {
  const parts: string[] = [];
  for (const [key, content] of Object.entries(filesMap)) {
    parts.push(`=== FILE: ${key} ===`);
    parts.push(content.trim());
  }
  return parts.join("\n\n");
}

// ── Clean header from XML ──

function cleanHeaderFromXml(xmlContent: string): string {
  const lines = xmlContent.split("\n");
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].trim().startsWith("<?xml")) {
      return lines.slice(i).join("\n");
    }
  }
  return xmlContent;
}

// ── DOCX ──

export function extractDocxContent(docxPath: string): string {
  const data = new Uint8Array(fs.readFileSync(docxPath));
  const zip = unzipSync(data);
  const docXml = zip["word/document.xml"];
  if (!docXml) throw new Error("Invalid DOCX: 'word/document.xml' not found");
  const xml = new TextDecoder().decode(docXml);
  return prettyPrintXml(xml);
}

export function saveDocxContent(docxPath: string, xmlContent: string): void {
  const data = new Uint8Array(fs.readFileSync(docxPath));
  const zip = unzipSync(data);

  if (!zip["word/document.xml"]) {
    throw new Error("Invalid DOCX: 'word/document.xml' not found");
  }

  const cleanXml = cleanHeaderFromXml(xmlContent);
  zip["word/document.xml"] = new TextEncoder().encode(cleanXml);

  const out = zipSync(zip);
  const tmpPath = docxPath + ".tmp";
  try {
    fs.writeFileSync(tmpPath, out);
    fs.renameSync(tmpPath, docxPath);
  } catch (e) {
    try { fs.unlinkSync(tmpPath); } catch {}
    throw e;
  }
}

// ── XLSX ──

export function extractXlsxContent(xlsxPath: string): string {
  const data = new Uint8Array(fs.readFileSync(xlsxPath));
  const zip = unzipSync(data);
  const result: Record<string, string> = {};

  // Extract worksheets
  for (const name of Object.keys(zip)) {
    if (name.startsWith("xl/worksheets/") && name.endsWith(".xml")) {
      const key = name.replace("xl/", "");
      result[key] = prettyPrintXml(new TextDecoder().decode(zip[name]));
    }
  }

  // Shared strings
  if (zip["xl/sharedStrings.xml"]) {
    result["sharedStrings.xml"] = prettyPrintXml(new TextDecoder().decode(zip["xl/sharedStrings.xml"]));
  }

  // Styles
  if (zip["xl/styles.xml"]) {
    result["styles.xml"] = prettyPrintXml(new TextDecoder().decode(zip["xl/styles.xml"]));
  }

  return dumpMultiFileContent(result);
}

export function saveXlsxContent(xlsxPath: string, multiFileContent: string): void {
  const xmlFiles = parseMultiFileContent(multiFileContent);
  const data = new Uint8Array(fs.readFileSync(xlsxPath));
  const zip = unzipSync(data);

  for (const [key, xmlContent] of Object.entries(xmlFiles)) {
    const fullPath = `xl/${key}`;
    const cleanXml = cleanHeaderFromXml(xmlContent);
    zip[fullPath] = new TextEncoder().encode(cleanXml);
  }

  const out = zipSync(zip);
  const tmpPath = xlsxPath + ".tmp";
  try {
    fs.writeFileSync(tmpPath, out);
    fs.renameSync(tmpPath, xlsxPath);
  } catch (e) {
    try { fs.unlinkSync(tmpPath); } catch {}
    throw e;
  }
}
